File: sdks/python/krymux/keys.py
"""Ed25519 identities: keypair generation, self-signed certificates, and the
canonical SPKI-sha256 fingerprint used for whitelisting and pinning.

The canonical identity of a peer is ``sha256:<hex>(sha256(SPKI DER))`` —
byte-identical to the Rust and Node reference implementations, so whitelists
and pins transfer 1:1 between implementations.
"""

from __future__ import annotations

import datetime
import hashlib
import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Union

from cryptography import x509
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.x509.oid import NameOID

PathLike = Union[str, Path]

#: DER SubjectPublicKeyInfo prefix for Ed25519 (RFC 8410) + 32-byte raw key.
ED25519_SPKI_PREFIX = bytes.fromhex("302a300506032b6570032100")

_CERT_VALID_FROM = datetime.datetime(2025, 1, 1, tzinfo=datetime.timezone.utc)
_CERT_VALID_TO = datetime.datetime(2049, 12, 31, tzinfo=datetime.timezone.utc)

_FP_RE = re.compile(r"^[0-9a-f]{64}$")


def spki_from_ed25519_raw(raw: bytes) -> bytes:
    """Build the DER SubjectPublicKeyInfo for a raw 32-byte Ed25519 key."""
    return ED25519_SPKI_PREFIX + raw


def fingerprint_of_spki(spki: bytes) -> str:
    """Canonical ``sha256:<hex>`` fingerprint of an SPKI DER blob."""
    return "sha256:" + hashlib.sha256(spki).hexdigest()


def _spki_der(cert: x509.Certificate) -> bytes:
    return cert.public_key().public_bytes(
        serialization.Encoding.DER, serialization.PublicFormat.SubjectPublicKeyInfo
    )


def fingerprint_of_cert(cert: x509.Certificate) -> str:
    """Canonical fingerprint of a certificate (sha256 over its SPKI DER)."""
    return fingerprint_of_spki(_spki_der(cert))


def fingerprint_of_cert_der(cert_der: bytes) -> str:
    """Canonical fingerprint of a DER-encoded certificate."""
    return fingerprint_of_cert(x509.load_der_x509_certificate(cert_der))


@dataclass
class Identity:
    """An Ed25519 identity: key pair, self-signed certificate, fingerprint."""

    key: ed25519.Ed25519PrivateKey
    cert: x509.Certificate
    fingerprint: str
    #: Source paths when loaded from disk (None when generated in memory).
    key_path: Path | None = None
    cert_path: Path | None = None
    _ssl_dir: Path | None = None

    @property
    def key_pem(self) -> bytes:
        return self.key.private_bytes(
            serialization.Encoding.PEM,
            serialization.PrivateFormat.PKCS8,
            serialization.NoEncryption(),
        )

    @property
    def cert_pem(self) -> bytes:
        return self.cert.public_bytes(serialization.Encoding.PEM)

    def ssl_pair(self) -> tuple[Path, Path]:
        """(cert_path, key_path) usable with ``ssl.SSLContext.load_cert_chain``.

        Always materializes the identity's own canonical PEM pair (clean
        PKCS#8 v1 key): source files written by the Rust CLI carry CRLF line
        endings and a PKCS#8 v2 layout that some OpenSSL PEM readers reject,
        so re-serialized copies are the portable form. Files are written once
        per Identity; call :meth:`cleanup` to remove them.
        """
        if self._ssl_dir is None:
            import tempfile

            self._ssl_dir = Path(tempfile.mkdtemp(prefix="krymux-identity-"))
            (self._ssl_dir / "identity.crt.pem").write_bytes(self.cert_pem)
            (self._ssl_dir / "identity.key.pem").write_bytes(self.key_pem)
        return self._ssl_dir / "identity.crt.pem", self._ssl_dir / "identity.key.pem"

    def cleanup(self) -> None:
        """Remove PEM files materialized by :meth:`ssl_pair` (if any)."""
        import shutil

        if self._ssl_dir is not None:
            shutil.rmtree(self._ssl_dir, ignore_errors=True)
            self._ssl_dir = None


def generate_identity(cn: str = "krymux") -> Identity:
    """Generate a fresh Ed25519 identity with a self-signed X.509 certificate.

    Mirrors the Rust reference: CN ``"<cn> <16 hex of fingerprint>"``, SAN
    ``krymux``, validity 2025-01-01 .. 2049-12-31.
    """
    key = ed25519.Ed25519PrivateKey.generate()
    spki = key.public_key().public_bytes(
        serialization.Encoding.DER, serialization.PublicFormat.SubjectPublicKeyInfo
    )
    fingerprint = fingerprint_of_spki(spki)
    short = fingerprint[len("sha256:") : len("sha256:") + 16]
    subject = x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, f"{cn} {short}")])
    cert = (
        x509.CertificateBuilder()
        .subject_name(subject)
        .issuer_name(subject)
        .public_key(key.public_key())
        .serial_number(x509.random_serial_number())
        .not_valid_before(_CERT_VALID_FROM)
        .not_valid_after(_CERT_VALID_TO)
        .add_extension(
            x509.SubjectAlternativeName([x509.DNSName("krymux")]), critical=False
        )
        .sign(key, algorithm=None)  # Ed25519: algorithm must be None
    )
    return Identity(key=key, cert=cert, fingerprint=fingerprint)


def _read_pem(data: bytes) -> bytes:
    """Normalize PEM bytes: the Rust CLI writes CRLF line endings, which
    OpenSSL tolerates but the cryptography parser does not."""
    return data.replace(b"\r\n", b"\n")


def _pem_body(data: bytes) -> bytes:
    import base64

    lines = [l for l in data.splitlines() if l and not l.startswith(b"-----")]
    return base64.b64decode(b"".join(lines))


def _der_children(buf: bytes) -> list[bytes]:
    """Split DER buffer into top-level TLV elements (bytes, each incl. tag)."""
    out: list[bytes] = []
    off = 0
    while off + 2 <= len(buf):
        length = buf[off + 1]
        header = 2
        if length & 0x80:
            n = length & 0x7F
            if n == 0 or n > 4 or off + 2 + n > len(buf):
                break
            length = int.from_bytes(buf[off + 2 : off + 2 + n], "big")
            header = 2 + n
        if off + header + length > len(buf):
            break
        out.append(buf[off : off + header + length])
        off += header + length
    return out


def _tlv_parts(elem: bytes) -> tuple[int, bytes]:
    """(tag, content) of one DER TLV element produced by ``_der_children``."""
    tag = elem[0]
    length = elem[1]
    header = 2
    if length & 0x80:
        n = length & 0x7F
        length = int.from_bytes(elem[2 : 2 + n], "big")
        header = 2 + n
    return tag, elem[header : header + length]


def _ed25519_from_pkcs8(pem: bytes) -> ed25519.Ed25519PrivateKey:
    """Tolerant PKCS#8 reader for Ed25519 keys.

    The Rust CLI (rcgen) emits PKCS#8 v2 (``version 1`` plus a ``[1]``
    public-key attribute); OpenSSL accepts it but the cryptography PKCS#8
    parser does not. Extract the 32-byte seed and rebuild the key directly.
    """
    der = _pem_body(pem)
    top = _der_children(der)
    if not top or top[0][0] != 0x30:
        raise ValueError("not a PKCS#8 private key")
    children = _der_children(_tlv_parts(top[0])[1])
    # children: [0] version INTEGER, [1] algorithm SEQUENCE, [2] privateKey OCTET STRING
    if len(children) < 3 or children[0][0] != 0x02 or children[2][0] != 0x04:
        raise ValueError("not a PKCS#8 private key")
    tag, alg = _tlv_parts(children[1])
    if b"\x06\x03\x2b\x65\x70" not in alg:  # OID 1.3.101.112 (Ed25519)
        raise ValueError("PKCS#8 key is not Ed25519")
    inner = _tlv_parts(children[2])[1]  # content of the outer OCTET STRING
    seed = inner[-32:]  # inner OCTET STRING (or raw) holding the 32-byte seed
    if len(seed) != 32:
        raise ValueError("cannot extract 32-byte Ed25519 seed")
    return ed25519.Ed25519PrivateKey.from_private_bytes(seed)


def load_identity(key_path: PathLike, cert_path: PathLike) -> Identity:
    """Load a key/certificate PEM pair and derive its fingerprint.

    Compatible with PEM files produced by ``krymux-tunnel keygen``.
    """
    key_path = Path(key_path)
    cert_path = Path(cert_path)
    key_pem = _read_pem(key_path.read_bytes())
    try:
        key = serialization.load_pem_private_key(key_pem, password=None)
    except ValueError:
        key = _ed25519_from_pkcs8(key_pem)
    if not isinstance(key, ed25519.Ed25519PrivateKey):
        raise ValueError(f"{key_path}: not an Ed25519 private key")
    cert = x509.load_pem_x509_certificate(_read_pem(cert_path.read_bytes()))
    # sanity: the key must match the certificate's public key
    kb = key.public_key().public_bytes(
        serialization.Encoding.DER, serialization.PublicFormat.SubjectPublicKeyInfo
    )
    cb = _spki_der(cert)
    if kb != cb:
        raise ValueError(f"{key_path} does not match {cert_path}")
    return Identity(
        key=key, cert=cert, fingerprint=fingerprint_of_cert(cert),
        key_path=key_path, cert_path=cert_path,
    )


def save_identity(directory: PathLike, identity: Identity, name: str) -> tuple[Path, Path]:
    """Persist an identity as ``<name>.key.pem`` / ``<name>.crt.pem`` plus an
    ``<name>.identity.json`` descriptor. Returns (key_path, cert_path)."""
    directory = Path(directory)
    directory.mkdir(parents=True, exist_ok=True)
    key_path = directory / f"{name}.key.pem"
    cert_path = directory / f"{name}.crt.pem"
    key_path.write_bytes(identity.key_pem)
    cert_path.write_bytes(identity.cert_pem)
    (directory / f"{name}.identity.json").write_text(
        json.dumps(
            {
                "name": name,
                "fingerprint": identity.fingerprint,
                "key": str(key_path),
                "cert": str(cert_path),
            },
            indent=2,
        )
        + "\n"
    )
    return key_path, cert_path


def normalize_fingerprint(text: str) -> str:
    """Accept flexible fingerprint inputs and return the canonical form.

    Accepted: ``sha256:<hex>``, bare 64-char hex, a path to a certificate PEM
    or a public-key (SPKI) PEM file, or PEM text starting with ``-----BEGIN``.
    """
    s = text.strip()
    if s.lower().startswith("sha256:"):
        h = re.sub(r"[:\s]", "", s[len("sha256:") :]).lower()
        if _FP_RE.match(h):
            return f"sha256:{h}"
        raise ValueError(f"bad sha256 fingerprint: {s}")
    if _FP_RE.match(s.lower()):
        return f"sha256:{s.lower()}"
    # A path (or pasted PEM) to a certificate or public key.
    if s.startswith("-----BEGIN") or Path(s).exists():
        data = _read_pem(s.encode() if s.startswith("-----BEGIN") else Path(s).read_bytes())
        try:
            cert = x509.load_pem_x509_certificate(data)
            return fingerprint_of_cert(cert)
        except ValueError:
            pass
        try:
            from cryptography.hazmat.primitives.serialization import load_pem_public_key

            pub = load_pem_public_key(data)
            spki = pub.public_bytes(
                serialization.Encoding.DER,
                serialization.PublicFormat.SubjectPublicKeyInfo,
            )
            return fingerprint_of_spki(spki)
        except Exception:
            pass
    raise ValueError(f"unrecognized fingerprint format: {text!r}")

File: crates/krymux/src/keys.rs
//! Ed25519 identities: keypair generation, self-signed certificates, and the
//! canonical fingerprint used for whitelisting and pinning.
//!
//! The canonical identity of a peer is `sha256(SPKI DER)` — byte-identical to
//! the Node implementation's fingerprint, so whitelists and pins transfer 1:1
//! between the two implementations.

use anyhow::{anyhow, bail, Context, Result};
use rcgen::{CertificateParams, DistinguishedName, DnType};
use rustls_pki_types::pem::PemObject;
use rustls_pki_types::{CertificateDer, PrivateKeyDer, SubjectPublicKeyInfoDer};
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};

/// DER: fixed SPKI prefix for Ed25519 (RFC 8410) + 32-byte raw key.
const ED25519_SPKI_PREFIX: [u8; 12] = [
    0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
];

/// Builds the DER SubjectPublicKeyInfo for a raw 32-byte Ed25519 public key.
pub fn spki_from_ed25519_raw(raw: &[u8]) -> Vec<u8> {
    let mut v = Vec::with_capacity(ED25519_SPKI_PREFIX.len() + raw.len());
    v.extend_from_slice(&ED25519_SPKI_PREFIX);
    v.extend_from_slice(raw);
    v
}

/// Computes the canonical `sha256:<hex>` fingerprint of an SPKI DER blob.
pub fn fingerprint_of_spki(spki: &[u8]) -> String {
    let h = Sha256::digest(spki);
    format!("sha256:{}", hex::encode(h))
}

// ---------------- minimal DER walker to extract the SPKI TLV ----------------

/// Parse one DER TLV header at `off`; returns (tag, content_len, header_len).
fn tlv_header(buf: &[u8], off: usize) -> Option<(u8, usize, usize)> {
    if off + 2 > buf.len() {
        return None;
    }
    let tag = buf[off];
    let first = buf[off + 1];
    let mut p = off + 2;
    let content_len = if first & 0x80 == 0 {
        first as usize
    } else {
        let n = (first & 0x7f) as usize;
        if n == 0 || n > 4 || p + n > buf.len() {
            return None;
        }
        let mut v = 0usize;
        for i in 0..n {
            v = (v << 8) | buf[p + i] as usize;
        }
        p += n;
        v
    };
    Some((tag, content_len, p - off))
}

/// Direct children of the TLV at [start, end) as (offset, total_len) pairs.
fn der_children(buf: &[u8], content_start: usize, content_end: usize) -> Vec<(usize, usize)> {
    let mut out = Vec::new();
    let mut off = content_start;
    while off < content_end {
        let Some((_, content_len, header_len)) = tlv_header(buf, off) else {
            break;
        };
        let total = header_len + content_len;
        if off + total > content_end {
            break;
        }
        out.push((off, total));
        off += total;
    }
    out
}

/// Extract the full SPKI TLV from an Ed25519 X.509 certificate DER.
pub fn extract_ed25519_spki(cert_der: &[u8]) -> Result<Vec<u8>> {
    // Certificate ::= SEQUENCE { tbsCertificate, signatureAlgorithm, signatureValue }
    let Some((_, clen, hlen)) = tlv_header(cert_der, 0) else {
        bail!("cert: bad outer header");
    };
    let top = der_children(cert_der, hlen, hlen + clen);
    let (tbs_off, tbs_total) = *top.first().ok_or_else(|| anyhow!("cert: no TBS element"))?;
    let tbs = &cert_der[tbs_off..tbs_off + tbs_total];
    let Some((_, tbs_clen, tbs_hlen)) = tlv_header(tbs, 0) else {
        bail!("cert: bad tbs header");
    };
    let children = der_children(tbs, tbs_hlen, tbs_hlen + tbs_clen);
    // children: [0]version?, serial, sig, issuer, validity, subject, spki, ...
    let mut idx = 0;
    if let Some((s, _)) = children.first() {
        if tbs[*s] == 0xa0 {
            idx = 1;
        }
    }
    let spki_pos = idx + 5;
    let (s, l) = *children
        .get(spki_pos)
        .ok_or_else(|| anyhow!("cert: no SPKI element"))?;
    let spki = tbs[s..s + l].to_vec();
    // sanity: must contain the Ed25519 OID 1.3.101.112
    if !spki.windows(5).any(|w| w == [0x06, 0x03, 0x2b, 0x65, 0x70]) {
        bail!("cert: SPKI is not Ed25519");
    }
    Ok(spki)
}

/// Computes the canonical fingerprint of an Ed25519 X.509 certificate by
/// extracting its SPKI.
pub fn fingerprint_of_cert(cert_der: &[u8]) -> Result<String> {
    let spki = extract_ed25519_spki(cert_der)?;
    Ok(fingerprint_of_spki(&spki))
}

// ---------------- identity generation / persistence ----------------

/// A freshly generated identity: key pair, self-signed certificate, and
/// fingerprint.
pub struct Identity {
    pub cert_der: CertificateDer<'static>,
    pub key_der: PrivateKeyDer<'static>,
    pub cert_pem: String,
    pub key_pem: String,
    pub fingerprint: String,
}

/// Generates a new Ed25519 identity; `cn` seeds the certificate common name.
pub fn generate_identity(cn: &str) -> Result<Identity> {
    // explicit Ed25519 — rcgen's generate() defaults to ECDSA P-256
    let key_pair =
        rcgen::KeyPair::generate_for(&rcgen::PKCS_ED25519).context("generate Ed25519 key")?;
    let fingerprint = {
        let spki = spki_from_ed25519_raw(key_pair.public_key_raw());
        fingerprint_of_spki(&spki)
    };
    let short: String = fingerprint
        .trim_start_matches("sha256:")
        .chars()
        .take(16)
        .collect();
    let mut params =
        CertificateParams::new(vec!["krymux".to_string()]).context("certificate params")?;
    let mut dn = DistinguishedName::new();
    dn.push(DnType::CommonName, format!("{} {}", cn, short));
    params.distinguished_name = dn;
    params.not_before = rcgen::date_time_ymd(2025, 1, 1);
    params.not_after = rcgen::date_time_ymd(2049, 12, 31);
    let cert = params
        .self_signed(&key_pair)
        .context("self-sign certificate")?;
    let cert_der = cert.der().clone();
    let cert_pem = cert.pem();
    let key_pem = key_pair.serialize_pem();
    let key_der = PrivateKeyDer::from_pem_slice(key_pem.as_bytes())
        .map_err(|e| anyhow!("reparse generated key: {e}"))?
        .clone_key();
    Ok(Identity {
        cert_der,
        key_der,
        cert_pem,
        key_pem,
        fingerprint,
    })
}

/// An identity loaded from PEM files on disk.
pub struct LoadedIdentity {
    pub cert_der: CertificateDer<'static>,
    pub key_der: PrivateKeyDer<'static>,
    pub fingerprint: String,
}

/// Loads a key/certificate PEM pair and derives its fingerprint.
pub fn load_identity(key_path: &Path, cert_path: &Path) -> Result<LoadedIdentity> {
    let key_der = PrivateKeyDer::from_pem_file(key_path)
        .with_context(|| format!("read key {}", key_path.display()))?
        .clone_key();
    let cert_der = CertificateDer::from_pem_file(cert_path)
        .with_context(|| format!("read cert {}", cert_path.display()))?
        .into_owned();
    let fingerprint = fingerprint_of_cert(cert_der.as_ref())?;
    Ok(LoadedIdentity {
        cert_der,
        key_der,
        fingerprint,
    })
}

/// Persists an identity under `dir` as `<name>.key.pem` / `<name>.crt.pem`
/// plus an `<name>.identity.json` descriptor (key file chmod 600 on Unix).
/// Returns the key and cert paths.
pub fn save_identity(dir: &Path, id: &Identity, name: &str) -> Result<(PathBuf, PathBuf)> {
    std::fs::create_dir_all(dir)?;
    let key_path = dir.join(format!("{}.key.pem", name));
    let cert_path = dir.join(format!("{}.crt.pem", name));
    std::fs::write(&key_path, &id.key_pem)?;
    std::fs::write(&cert_path, &id.cert_pem)?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&key_path, std::fs::Permissions::from_mode(0o600))
            .with_context(|| format!("chmod 600 {}", key_path.display()))?;
    }
    let meta = serde_json::json!({
        "name": name,
        "fingerprint": id.fingerprint,
        "key": key_path.display().to_string(),
        "cert": cert_path.display().to_string(),
    });
    std::fs::write(
        dir.join(format!("{}.identity.json", name)),
        serde_json::to_string_pretty(&meta)?,
    )?;
    Ok((key_path, cert_path))
}

/// Accept flexible fingerprint inputs: `sha256:<hex>`, bare hex (64), or a path
/// to a certificate / public key PEM file.
pub fn normalize_fingerprint(input: &str) -> Result<String> {
    let s = input.trim();
    if let Some(hexpart) = s.strip_prefix("sha256:") {
        let h: String = hexpart.chars().filter(|c| *c != ':' && *c != ' ').collect();
        let h = h.to_lowercase();
        if h.len() == 64 && h.chars().all(|c| c.is_ascii_hexdigit()) {
            return Ok(format!("sha256:{}", h));
        }
        bail!("bad sha256 fingerprint: {}", s);
    }
    if s.len() == 64 && s.chars().all(|c| c.is_ascii_hexdigit()) {
        return Ok(format!("sha256:{}", s.to_lowercase()));
    }
    let path = Path::new(s);
    if s.starts_with("-----BEGIN") || path.exists() {
        if let Ok(cert) = CertificateDer::from_pem_file(path) {
            return fingerprint_of_cert(cert.as_ref());
        }
        if let Ok(spki) = SubjectPublicKeyInfoDer::from_pem_file(path) {
            return Ok(fingerprint_of_spki(spki.as_ref()));
        }
    }
    bail!("unrecognized fingerprint format");
}
